circuit-construction/src/lib.rs

```diff
@@ -233,6 +233,61 @@ pub trait Cs<F: FftField + PrimeField> {
         });
     }
 
+    fn and(&mut self, x1: Var<F>, x2: Var<F>) -> Var<F> {
```

Comment:
Indeed, they are probably better to be encapsulated in a `Boolean` type, otherwise it can be confusing with the field elements in practice. For now, I think I can just work on adding arithmetic functions for typical field elements by following the current structure in `Cs` trait.

Meanwhile, I am thinking maybe the `Cs` trait can be splited into different traits according to the operation categories, such as `field` (`add/sub/mul/eq/gt/lt...`), `boolean` (`and/or/not...`) and `curve points` (`add_group/double/endo...`). 

Under current structure, the value type is _locked_ once a `WitnessGenerator` or `System` is created, and all the operations are assumed to use the value type in a life cycle to keep track the gate rows/cofficients.

If the `Cs.var` can be created based on the different types, and the created `vars` can be still kept tracked by the `WitnessGenerator` or `System`, then the `vars` might become extensible to do calculation with different types while having rust does the type checking. This is just a thought that I am still brewing 😋

Reply:
I think it'd be a good idea to try to create a `Bool(ean)` type that would encapsulate a `Var`, or implement a trait that `Var` also implements (that's what [arkworks r1cs std does](https://github.com/arkworks-rs/r1cs-std/blob/master/src/bits/boolean.rs#L164) by implementing R1CSVar and AllocVar). But maybe this should wait for something like https://github.com/o1-labs/proof-systems/pull/596